enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,383 @@
+manifest introduced in the imagestream which will contain the image matching that of the control plane architecture. For imports into the image
+registry, a manifestlist containing a single image matching that of the control plane architecture will be imported (if the option to preserve
+manifestlist is chosen. if not it would be a single manifest image). This ensures that the image registry behavior remains as is and current CI
+tooling is not affected. At the same time, for non control plane architecture workers, while the image  registry images would not work, they can
+be overriden with the manifest listed imagestream. Eventually the oc tooling would also be upgraded to point to the manifest listed image from
+source for non control plane architecture workers.
+
+### OLM
+
+OLM would need to make sure that it deploys operators only to nodes whose architecture matches the architectures on which the operator is supported.
+Today, OLM has a subscription object which is used to tell it how to deploy an operator. It has support for the nodeselector field but this would
+not be sufficient to specify filtering on multiple architectures. To this end, it would have to support the podAffinity field which uses the
+nodeselectorterms to filter on multiple values.
+
+Instead of having OLM dictate the operator deployment on specific nodes, the individual operators themselves would specify which architectures
+they support through a nodeselector object which OLM will then provide the ability for the user to override if necessary. This information should
+also be added to the operands and potentially be made configurable by the user requesting the operand. The operator-sdk framework could also be
+potentially leveraged to support the generation of this supported architecture list when creating operators.
+
+### Machine Config Operator
+
+No changes are needed in the MCO for Phase 1 as the machine-os-content image would be a manifest listed image and the machine config daemon would
+extract the relevant architecture's machine-os-content based on the node it runs on. In the future, architecture specific configurations should
+not be required, instead they need to be templatized and be generalized for certain items, for example kubelet system reserved memory could scale
+based on page size.
+
+### Machine API
+
+- As of today, machinesets can be mutated to change the instance type and image id and this could potentially result in inconsistency between existing
+machinesets and newly created machinesets. If the instance type and image is changed to one of a different architecture, there would be workers of
+different architectures with the same machineset definition. While this may not be a problem, it doesn't make for a good user experience. Hence,
+machineset specs will not allowed to be mutated. Anytime a change is required, a new machineset needs to be created and the old one needs to be scaled
```

Comment:
What happens today in a homogeneous cluster if the instance type is modified to one that uses a different architecture? Do we already block users from editing that field or is this a big API change?

Reply:
today we allow users to change instance types. what happens is - the  newer machines created will be of the new instance type while the older machines will persist, ending up in a mismatch. upon discussion with the cloud team the idea was to block this change and ask users to create a new machineset instead of changing existing ones.

Reply:
Is there an easy way that we can do this across platforms? While we have admission webhooks already for some platforms, changing the behaviour to make part of the MachineSet providerSpec immutable where it wasn't before sounds like a breaking change.

Are we going to need to enumerate amd64 and arm64 instance types and make sure that if your MachineSet was created as one, it cannot be updated to the other? 

Note, in our existing webhooks, we don't have the old spec when validating, so that will need to be changed

Have we weighed up the complexity of making this immutable vs the user shooting themselves in the foot factor? Do we need a technical solution for this?

Reply:
yes part of the problem today is that there is no way to differentiate arches today. i remember in the discussion, it was agreed upon to make it immutable, but not sure the details were hashed out after that. @elmiko might have further ideas.

Reply:
i would assume this would mean making instance type immutable, but not sure if that would apply to the spec as a whole to make it immutable in it's entirety.

Reply:
We already today have instructions in our product docs (those for resizing the control plane) that talk about changing the instance type of a Machine when you resize it out of band, we need to make sure we don't break scenarios like this that are already documented.

Reply:
reading back here, i feel like if we decide to start doing validations on the fields of the provider spec we will get into some serious complexity quickly. making them immutable will probably shock some users who are accustomed to having the ability to change the provider spec, but imo will lead us to a better place in terms of best cluster practices. from a certain standpoint creating a new machineset and updating the provider spec on a current one aren't much different. the user will still need to scale up new instances to get the new provider changes.

Reply:
have we decided whether this solution is worth pursuing or leaving as is? @JoelSpeed @elmiko ?